llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -5209,6 +5209,26 @@ PerformBUILD_VECTORCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI) {
   return DAG.getNode(ISD::BITCAST, DL, VT, PRMT);
 }
 
+static SDValue combineADDRSPACECAST(SDNode *N,
+                                    TargetLowering::DAGCombinerInfo &DCI) {
+  auto *ASCN1 = cast<AddrSpaceCastSDNode>(N);
+
+  if (auto *ASCN2 = dyn_cast<AddrSpaceCastSDNode>(ASCN1->getOperand(0))) {
+    assert(ASCN2->getDestAddressSpace() == ASCN1->getSrcAddressSpace());
+
+    // Fold asc[B -> A](asc[A -> B](x)) -> x
+    if (ASCN1->getDestAddressSpace() == ASCN2->getSrcAddressSpace())
+      return ASCN2->getOperand(0);
```

Comment:
I think we intentionally insert `ASC(generic->global->generic)` to allow tracking specific AS the pointer actually points to, so we can lower loads/stores using AS-specific instruction. Collapsing it all back to `generic` will lose the hint, if it kicks in before we've lowered loads/stores.

Reply:
We do insert these casts, but they really only serve as a hint to InferAddressSpaces, by the time we get to SelectionDAG there won't be any future passes that would use this hint.